src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -598,8 +616,42 @@ ss::future<cluster::init_tm_tx_reply> tx_gateway_frontend::init_tm_tx_locally(
           return ss::with_gate(
             self._gate,
             [tx_id, transaction_timeout_ms, timeout, expected_pid, &self] {
-                return self.do_init_tm_tx(
-                  tx_id, transaction_timeout_ms, timeout, expected_pid);
+                return self.with_stm(
+                  [tx_id, transaction_timeout_ms, timeout, expected_pid, &self](
+                    checked<ss::shared_ptr<tm_stm>, tx_errc> r) {
+                      if (!r) {
+                          return ss::make_ready_future<
+                            cluster::init_tm_tx_reply>(
+                            cluster::init_tm_tx_reply{r.error()});
+                      }
+                      auto stm = r.value();
+                      return stm->read_lock().then(
+                        [&self,
+                         stm,
+                         tx_id,
+                         transaction_timeout_ms,
+                         expected_pid,
+                         timeout](ss::basic_rwlock<>::holder unit) {
+                            return with(
+                                     stm,
+                                     tx_id,
+                                     "init_tm_tx",
+                                     [&self,
+                                      stm,
+                                      tx_id,
+                                      transaction_timeout_ms,
+                                      expected_pid,
+                                      timeout]() {
+                                         return self.do_init_tm_tx(
+                                           stm,
+                                           tx_id,
+                                           transaction_timeout_ms,
+                                           timeout,
+                                           expected_pid);
+                                     })
+                              .finally([u = std::move(unit)] {});
```

Comment:
since `unit` holds a reference to _state_lock doesn't `stm` need to also be kept alive?

Reply:
discussed offline, by the code composition stm is alive when the stm's gate is open so extending isn't needed; also we pass stm into the fiber by value so extending isn't needed